beginner_source/aws_distributed_training_tutorial.py

```diff
@@ -0,0 +1,675 @@
+# best performing model then save it at the end of training (see
+# `here <https://github.com/pytorch/examples/blob/master/imagenet/main.py#L184>`__).
+# 
+
+best_prec1 = 0
+
+for epoch in range(num_epochs):
+    # Set epoch count for DistributedSampler
+    train_sampler.set_epoch(epoch)
+    
+    # Adjust learning rate according to schedule
+    adjust_learning_rate(starting_lr, optimizer, epoch)
+
+    # train for one epoch
+    print("\nBegin Training Epoch {}".format(epoch+1))
+    train(train_loader, model, criterion, optimizer, epoch)
+
+    # evaluate on validation set
+    print("Begin Validation @ Epoch {}".format(epoch+1))
+    prec1 = validate(val_loader, model, criterion)
+
+    # remember best prec@1 and save checkpoint if desired
+    #is_best = prec1 > best_prec1
+    best_prec1 = max(prec1, best_prec1)
+
+    print("Epoch Summary: ")
+    print("\tEpoch Accuracy: {}".format(prec1))
+    print("\tBest Accuracy: {}".format(best_prec1))
+
+
+######################################################################
+# Running the Code
+# ----------------
+# 
+# Unlike most of the other PyTorch tutorials, this code may not be run
+# directly out of this notebook. To run, download the .py version of this
+# file (or convert it using
+# `this <https://gist.github.com/chsasank/7218ca16f8d022e02a9c0deb94a310fe>`__)
+# and upload a copy to both nodes. The astute reader would have noticed
+# that we hardcoded the **node0-privateIP** and :math:`world\_size=4` but
+# input the *rank* and *local\_rank* inputs as arg[1] and arg[2] command
+# line arguments, respectively. Once uploaded, open two ssh terminals into
+# each node.
+# 
+# -  On the first terminal for node0, run ``$ python main.py 0 0``
+# 
+# -  On the second terminal for node0 run ``$ python main.py 1 1``
+# 
+# -  On the first terminal for node1, run ``$ python main.py 2 0``
+# 
+# -  On the second terminal for node1 run ``$ python main.py 3 1``
+# 
+# The programs will start and wait after printing "Initialize Model..."
+# for all four processes to join the process group. Notice the first
+# argument is not repeated as this is the unique global rank of the
+# process. The second argument is repeated as that is the local rank of
+# the process running on the node. If you run ``nvidia-smi`` on each node,
+# you will see two processes on each node, one running on GPU0 and one on
+# GPU1.
+# 
+# We have now completed the distributed training example! Hopefully you
+# can see how you would use this tutorial to help train your own models on
+# your own datasets, even if you are not using the exact same distributed
+# envrionment. If you are using AWS, don't forget to **SHUT DOWN YOUR
+# NODES** if you are not using them or you may find an uncomfortably large
+# bill at the end of the month.
+# 
+# **Where to go next**
+# 
+# -  Check out the `launcher
+#    utility <https://pytorch.org/docs/stable/distributed.html#launch-utility>`__
+#    for a different way of kicking off the run
+# 
```

Comment:
If you use python 3.4 or above, check out the ImageNet example in pytorch example. This will use `torch.multiprocessing.spawn`. 

Please also mention about this for another way to launch multiprocesses. Please also take a look at the latest pytorch example image net code, which has this implemented